sumtype on map value causes compile error

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: macos, macOS, 12.1, 21C52
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.0.0 (clang-1300.0.29.30)

getwd: /Users/hunam/Documents/dev/go2v/letsgetphysical
vmodules: /Users/hunam/.vmodules
vroot: /Applications/v
vexe: /Applications/v/v
vexe mtime: 2022-01-26 10:07:05
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 36c9ed6.0979723

Git version: git version 2.32.0 (Apple Git-132)
Git vroot status: weekly.2022.04-20-g09797236 (4 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-unknown-unknown de82a130
```

**What did you do?**
```v
type Child = Tree | string

struct Tree {
	name  string
	child map[string]Child
}

fn main() {
	tree := Tree{
		name: 'foo'
		child: {
			'a': 'value'
		}
	}
}
```
**What did you expect to see?**
no compilation error
 
**What did you see instead?**
```v
/tmp/v_501/report.6740966653263284261.tmp.c:1986:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/report.6740966653263284261.tmp.c:11037:242: error: initializing 'main__Tree *' (aka 'struct main__Tree *') with an expression of incompatible type 'string' (aka 'struct string')
        main__Tree tree = (main__Tree){.name = _SLIT("foo"),.child = new_map_init(&map_hash_string, &map_eq_string, &map_clone_string, &map_free_string, 1, sizeof(string), sizeof(main__Child), _MOV((string[1]){_SLIT("a"), }), _MOV((main__Child[1]){_SLIT("value"), })),};
                                                                                                                                                                                                                                                        ^~~~~~~~~~~~~~
/tmp/v_501/report.6740966653263284261.tmp.c:158:18: note: expanded from macro '_SLIT'
#define _SLIT(s) ((string){.str=(byteptr)("" s), .len=(sizeof(s)-1), .is_lit=1})
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
```

--- a/cmd/tools/vast/vast.v
+++ b/cmd/tools/vast/vast.v
@@ -1589,6 +1589,7 @@ fn (t Tree) map_init(node ast.MapInit) &Node {
 	obj.add_terse('value_type', t.type_node(node.value_type))
 	obj.add_terse('keys', t.array_node_expr(node.keys))
 	obj.add_terse('vals', t.array_node_expr(node.vals))
+	obj.add_terse('val_types', t.array_node_type(node.val_types))
 	obj.add('comments', t.two_dimension_comment(node.comments))
 	obj.add('pre_cmnts', t.array_node_comment(node.pre_cmnts))
 	obj.add('pos', t.pos(node.pos))

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -1200,6 +1200,7 @@ pub:
 pub mut:
 	keys       []Expr
 	vals       []Expr
+	val_types  []Type
 	typ        Type
 	key_type   Type
 	value_type Type

--- a/vlib/v/checker/containers.v
+++ b/vlib/v/checker/containers.v
@@ -247,6 +247,7 @@ pub fn (mut c Checker) map_init(mut node ast.MapInit) ast.Type {
 			if node.vals[0].is_auto_deref_var() {
 				val0_type = val0_type.deref()
 			}
+			node.val_types << val0_type
 		}
 		mut same_key_type := true
 		for i, key in node.keys {
@@ -258,6 +259,7 @@ pub fn (mut c Checker) map_init(mut node ast.MapInit) ast.Type {
 			key_type := c.expr(key)
 			c.expected_type = val0_type
 			val_type := c.expr(val)
+			node.val_types << val_type
 			if !c.check_types(key_type, key0_type) || (i == 0 && key_type.is_number()
 				&& key0_type.is_number() && key0_type != ast.mktyp(key_type)) {
 				msg := c.expected_msg(key_type, key0_type)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4167,9 +4167,9 @@ fn (mut g Gen) map_init(node ast.MapInit) {
 	unwrap_val_typ := g.unwrap_generic(node.value_type)
 	key_typ_str := g.typ(unwrap_key_typ)
 	value_typ_str := g.typ(unwrap_val_typ)
-	value_typ := g.table.sym(unwrap_val_typ)
-	key_typ := g.table.final_sym(unwrap_key_typ)
-	hash_fn, key_eq_fn, clone_fn, free_fn := g.map_fn_ptrs(key_typ)
+	value_sym := g.table.sym(unwrap_val_typ)
+	key_sym := g.table.final_sym(unwrap_key_typ)
+	hash_fn, key_eq_fn, clone_fn, free_fn := g.map_fn_ptrs(key_sym)
 	size := node.vals.len
 	mut shared_styp := '' // only needed for shared &[]{...}
 	mut styp := ''
@@ -4198,7 +4198,7 @@ fn (mut g Gen) map_init(node ast.MapInit) {
 		}
 	}
 	if size > 0 {
-		if value_typ.kind == .function {
+		if value_sym.kind == .function {
 			g.write('new_map_init${noscan}($hash_fn, $key_eq_fn, $clone_fn, $free_fn, $size, sizeof($key_typ_str), sizeof(voidptr), _MOV(($key_typ_str[$size]){')
 		} else {
 			g.write('new_map_init${noscan}($hash_fn, $key_eq_fn, $clone_fn, $free_fn, $size, sizeof($key_typ_str), sizeof($value_typ_str), _MOV(($key_typ_str[$size]){')
@@ -4207,16 +4207,20 @@ fn (mut g Gen) map_init(node ast.MapInit) {
 			g.expr(expr)
 			g.write(', ')
 		}
-		if value_typ.kind == .function {
+		if value_sym.kind == .function {
 			g.write('}), _MOV((voidptr[$size]){')
 		} else {
 			g.write('}), _MOV(($value_typ_str[$size]){')
 		}
-		for expr in node.vals {
+		for i, expr in node.vals {
 			if expr.is_auto_deref_var() {
 				g.write('*')
 			}
-			g.expr(expr)
+			if value_sym.kind == .sum_type {
+				g.expr_with_cast(expr, node.val_types[i], unwrap_val_typ)
+			} else {
+				g.expr(expr)
+			}
 			g.write(', ')
 		}
 		g.write('}))')

--- a/vlib/v/tests/map_sumtype_value_init_test.v
+++ b/vlib/v/tests/map_sumtype_value_init_test.v
@@ -1,0 +1,18 @@
+type Child = Tree | string
+
+struct Tree {
+	name  string
+	child map[string]Child
+}
+
+fn test_map_sumtype_value_init() {
+	tree := Tree{
+		name: 'foo'
+		child: {
+			'a': 'value'
+		}
+	}
+	println(tree)
+	ret := tree.child['a'] or { Child('') }
+	assert ret == Child('value')
+}

